enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,229 @@
+When a Device Administrator triggers a software or operating system upgrade
+on Edge Devices based on `rpm-ostree`, the changes do not take place until the
+next system reboot.
+
+**End User** is a human user operating an Edge Device. End users continue their
+work with the device running the previous software version until the reboot.
+
+Once the system is rebooted implicitly (i.e. `cron` schedule) or explicitly
+by a Device Administrator, 
+the [greenboot-healthcheck.service](https://github.com/fedora-iot/greenboot#how-does-it-work)
+runs to verify the system validity.
+
+If the system validity procedures are successful, the version of the software
+is declared valid for end users. Otherwise, `greenboot` enters a system
+verification loop by repeating reboots up-to `GREENBOOT_MAX_BOOT_ATTEMPTS` times
+before attempting to rollback to the previous version of the software.
+
+[![Mermaid Upgrade Flow](https://mermaid.ink/img/pako:eNpV0cFuwjAMANBfsXIdHHbtYVJRVzRpTBMMLk0PpnEhIk2iNGVDbf99oammkVMUP1uO3bPKCGIJOzm0Z_jKuIZw0mJnav-NjuD6UcJy-QLD38veBixogFWxpaMxHj5JC6lPZUxezf7WemogkgGy_oBKCvQE_2o_PY8xKYtJkxngtXgw5YN509eo8uJATtayQi-Nhndj7CzzKPcWvIEN_sxdtEP2EE-VghylIgGpD83aIGBdbI1SR6wuc7H1hFOu2YI15BqUIgysvwc582dqiLMkXAW6C2dcj8Fh583upiuWeNfRgnX2_vNMYhhdw5IaVRteSUhv3CZuYFrE-AvrnYIR?type=png)](https://mermaid.live/edit#pako:eNpV0cFuwjAMANBfsXIdHHbtYVJRVzRpTBMMLk0PpnEhIk2iNGVDbf99oammkVMUP1uO3bPKCGIJOzm0Z_jKuIZw0mJnav-NjuD6UcJy-QLD38veBixogFWxpaMxHj5JC6lPZUxezf7WemogkgGy_oBKCvQE_2o_PY8xKYtJkxngtXgw5YN509eo8uJATtayQi-Nhndj7CzzKPcWvIEN_sxdtEP2EE-VghylIgGpD83aIGBdbI1SR6wuc7H1hFOu2YI15BqUIgysvwc582dqiLMkXAW6C2dcj8Fh583upiuWeNfRgnX2_vNMYhhdw5IaVRteSUhv3CZuYFrE-AvrnYIR)
+
+### API Extensions
+
+None
+
+### Implementation Details
+
+The MicroShift core RPM package will include the `microshift-greenboot-required.sh`
+health check script to validate that all the required MicroShift services are
+up and running. There will be no explicit dependency on the `greenboot` RPM.
+The script will be installed into the `/etc/greenboot/check/required.d` directory
+and it will not be executed during the system boot in case the `greenboot` package
+is not present.
+
+> If other scripts are present in the `/etc/greenboot/check/required.d` directory,
+> they will be executed by `greenboot` in the alphanumeric order.
+
+The script performs the following sanity checks:
+* Exit if the current user is not `root`
+* Exit if MicroShift service is not enabled
+* Wait for the MicroShift service to become active
+
+If the `greenboot-default-health-check` RPM subpackage is installed, it already
+includes [health check scripts](https://github.com/fedora-iot/greenboot#health-checks-included-with-subpackage-greenboot-default-health-checks)
+verifying that DNS and `ostree` services can be accessed.
+
+The health check script will implement validations for the core MicroShift state
+and its services running as pods in the following namespaces:
```

Comment:
By "core MicroShift state" do you mean using the healthcheck API endpoints for the API server, etc.?

Reply:
MicroShift does its own health checks, so the fact that the service is active seems to answer that question.
Do we need to add a special check for the healthcheck APIs?

Reply:
If systemd doesn't report the service as active until MicroShift's healthz endpoint would report it as healthy, then I think it's OK to rely just on systemd. If those 2 statuses are not linked then we need to check both.

Reply:
I added the Kube API health endpoint check